dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/programrule/engine/RuleActionScheduleMessageImplementer.java

```diff
@@ -0,0 +1,151 @@
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import org.hisp.dhis.common.IdentifiableObjectStore;
+import org.hisp.dhis.notification.logging.ExternalNotificationLogEntry;
+import org.hisp.dhis.notification.logging.NotificationLoggingService;
+import org.hisp.dhis.notification.logging.NotificationTriggerEvent;
+import org.hisp.dhis.program.ProgramInstance;
+import org.hisp.dhis.program.ProgramStageInstance;
+import org.hisp.dhis.program.notification.ProgramNotificationInstance;
+import org.hisp.dhis.program.notification.ProgramNotificationTemplate;
+import org.hisp.dhis.rules.models.RuleAction;
+import org.hisp.dhis.rules.models.RuleActionScheduleMessage;
+import org.hisp.dhis.rules.models.RuleEffect;
+import org.hisp.dhis.system.util.DateUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+
+/**
+ * @Author Zubair Asghar.
+ */
+public class RuleActionScheduleMessageImplementer extends BaseRuleActionImplementer
+{
+    // -------------------------------------------------------------------------
+    // Dependencies
+    // -------------------------------------------------------------------------
+
+    @Autowired
+    private NotificationLoggingService notificationLoggingService;
+
+    @Autowired @Qualifier( "org.hisp.dhis.program.notification.ProgramNotificationInstanceStore" )
+    private IdentifiableObjectStore<ProgramNotificationInstance> programNotificationInstanceStore;
+
+    @Override
+    public boolean accept( RuleAction ruleAction )
+    {
+        return ruleAction != null && ruleAction instanceof RuleActionScheduleMessage;
+    }
+
+    @Override
+    public void implement( RuleEffect ruleEffect, ProgramInstance programInstance )
+    {
+        if ( !validate( ruleEffect, programInstance ) )
+        {
+            return;
+        }
+
+        ProgramNotificationTemplate template = getNotificationTemplate( ruleEffect.ruleAction() );
+
+        String key = generateKey( template, programInstance );
+
+        String date = ruleEffect.data();
+
+        if ( !isDateValid( date ) )
+        {
+            return;
+        }
```

Comment:
The code for the schedule implementer and sendmessage implementer seems to be the same down to this point. Perhaps we can pull more into the abstract base class here.